### PR TITLE
Sockets - Amy

### PR DESCRIPTION
**Pull request by aphunk**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? |  It was really challenging to decide which methods lived where; for example, I set up the `reserve room` method in the `concierge` class, but as the project went on, I realized it might have been better to live in the `reservation` class.
What was a design decision you made that changed over time over the project? | The `DateRange` class went through many changes throughout the project. It sometimes felt necessary to have as a class, and other times it made more sense as just a method, but ultimately, it is a class itself.
What was a concept you gained clarity on, or a learning that you'd like to share? | I gained more clarity on the importance of limiting dependencies--with so many relationships in the hotel project, I tried to do as much dependence injection as I could, where I could.
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | The test for viewing available rooms. It's nominal because it's testing that the method behaves in the way it's supposed to by returning a list of available rooms for the given date
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | The test for when no rooms are available for the given date. It's an edge case because it is an extreme case where you'd expect there to be an available room on a given night.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I felt very good about the pseudocode, but to be honest, it fell apart a bit going from writing tests to then the code. Writing tests in general is a skill that I need a lot of work on, and so I found it difficult to do that prior to writing the code. Looking back, I can see how it could have helped the code if I had stuck to a process of writing tests before code.

### CODE REVIEW

File: lib/date_range.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  end
+  
+  def include_date_range?(date_range)
+    @start_date < date_range.start_date && @end_date >= date_range.end_date
+  end
+
+  def overlap_date_range?(date_range)
+    if @start_date <= date_range.start_date && @end_date > date_range.start_date
+      return true
+    end
```
Comment (droberts-sea) [inline lib/date_range.rb:14]: Good work on this tricky logic.

File: lib/room.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
+module Hotel
+  class Room
+    attr_accessor :status
+    attr_reader :room_number, :reservations
+    
+      def initialize(room_number:, status: :AVAILABLE, reservations: [])
+        @room_number = room_number
+        @status = status
```
Comment (droberts-sea) [inline lib/room.rb]: The room status is only ever `:AVAILABLE` - you never set it to `:UNAVAILABLE`. Maybe you could remove this instance variable.

File: spec/reservation_spec.rb [hunk 3]
```diff
@@ -0,0 +1,77 @@
+require "spec_helper"
+require "date"
+
```
Comment (droberts-sea) [inline spec/reservation_spec.rb:3]: Line 1 should be `require_relative` (this prevented your tests from running)

File: lib/room.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+      end
+    
+    
+      def add_reservation(reservation)
+        @reservations << reservation
+      end
+      
+      def is_available?(date_range)
+        return false if @reservations.each { |res| res.overlap_date_range?(date_range) } 
+
```
Comment (droberts-sea) [inline lib/room.rb]: This is very close, but I don't think it's quite doing what you expect. In fact, this method will always return `false`!

```ruby
$ pry -r ./lib/concierge.rb
pry> room = Hotel::Room.new(room_number: 1)
pry> dr = DateRange.new(start_date: 'March 17th, 2019', end_date: 'March 20th, 2019')
pry> room.is_available?(dr) # no reservations yet
=> false
```

The reason is that the `.each` method returns a reference to the original array (`@reservations` in this case), which is always truthy. Instead, you might do something like this:

```ruby
return !@reservations.any { |res| res.overlap_date_range?(date_range) }
```

File: lib/concierge.rb [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        @all_rooms << Hotel::Room.new(room_number: i + 1)
+      end
+      @reservations = []
+    end
+        
+        
+    def reserve_room(start_date, end_date)
+      available_room = @all_rooms.find do |room| 
+        room.status == :AVAILABLE
+      end
```
Comment (droberts-sea) [inline lib/concierge.rb:23]: Instead of checking `room.status` here, you might want to call `room.is_available?`.

File: lib/reservation.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
+require "date"
+
+require_relative "room"
+require_relative "date_range"
+
+module Hotel
+  class Reservation
+    attr_reader :id, :room, :start_date, :end_date, :total_cost, :date_range
+
```
Comment (droberts-sea) [inline lib/reservation.rb]: Since `total_cost` isn't an instance variable, you don't need to have a `attr_reader` for it.

File: lib/date_range.rb [hunk 7]
```diff
@@ -0,0 +1,35 @@
+require 'date'
+
+class DateRange
+  attr_accessor :start_date, :end_date
+  
+  def initialize(start_date:, end_date:)
+    @start_date = Date.parse(start_date)
+    @end_date = Date.parse(end_date)
+  end
```
Comment (droberts-sea) [inline lib/date_range.rb]: This would be a great place to put the check for whether the end date is after the start date.

File: lib/reservation.rb [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+module Hotel
+  class Reservation
+    attr_reader :id, :room, :start_date, :end_date, :total_cost, :date_range
+    
+    def initialize(id:, room:, start_date:, end_date:, date_range: nil)
+      @id = id
+      @room = room
+      @start_date = start_date
+      @end_date = end_date
+      @date_range = date_range
```
Comment (droberts-sea) [inline lib/reservation.rb]: This implies that it would be OK to create a `Reservation` with either a start and end date, or a date range. I like this idea, but your implementation doesn't quite get there. In particular the `date_range` method below will override the generated `attr_reader`, so that anything stored in `@date_range` is invisible.

I think my recommendation would be to have something like this in the constructor:

```ruby
def initialize(id:, room:, start_date: nil, end_date: nil, date_range: nil)
  @id = id
  @room = room

  if date_range
    @date_range = date_range

  else
    unless start_date && end_date
      raise ArgumentError, "one of start + end date or date range is required"
    end
    # Constructor for DateRange checks range validity
    @date_range = DateRange.new(start_date: start_date, end_date: end_date)
  end
end
```

File: lib/date_range.rb [hunk 9]
```diff
@@ -0,0 +1,35 @@
+require 'date'
+
+class DateRange
+  attr_accessor :start_date, :end_date
+  
+  def initialize(start_date:, end_date:)
```
Comment (droberts-sea) [inline lib/date_range.rb]: I really love this design decision to put `DateRange` into its own class. Keeping all of this code in one place makes the rest of the project much more readable, and makes it very obvious where I the reader need to go for any date-related calculations.

File: lib/date_range.rb [hunk 10]
```diff
@@ -0,0 +1,35 @@
+require 'date'
+
+class DateRange
+  attr_accessor :start_date, :end_date
+
```
Comment (droberts-sea) [inline lib/date_range.rb]: Should this be inside the `Hotel` module?

File: lib/date_range.rb (hunk 7 above)
Comment (droberts-sea) [reply lib/date_range.rb]: I like the dedication to keyword arguments throughout this project - good work experimenting with this new concept.

This is a style thing, and you're welcome to disagree, but for this specific method you might not need them. Keyword arguments are useful when the order of the arguments isn't obvious, but I feel that putting `start_date` before `end_date` would be pretty natural.

File: lib/concierge.rb [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    
+    
+    def view_reservations_by_date(date)
+      result = @reservations.filter {|res| res.include_date?(date)}
+        return result
+    end
+
+    def self.see_all_rooms
+      return @all_rooms
+    end
```
Comment (droberts-sea) [inline lib/concierge.rb:38]: I'm not sure what this method is doing. Having a class method access an instance variable doesn't make sense, and you already have an `attr_reader` for `all_rooms`.

File: lib/concierge.rb [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+      available_rooms = []
+      rooms_w_res = []
+      @all_rooms.map do |room|
+        if room.reservations == []
+          available_rooms << room
+        else
+          rooms_w_res << room
+          
+          rooms_w_res.each do |room|
+            room.reservations.each do |res|
```
Comment (droberts-sea) [inline lib/concierge.rb]: A few comments on this code:

- You call `map` on line 52, but don't do anything with the result. `map` is a specific pattern, but it doesn't seem to match what this code is doing. I think that `each` might be a better choice here.
- I think you want to end the previous loop before starting the next one on line 58. If you don't do so, this code will give the wrong result if there's more than one reservation for that date range.
- I also don't think you need to treat rooms without reservations separately. On line 59 you loop through the reservations for that room, which will work fine (executing 0 times) if there are no reservations, giving you the same result.

However, I don't know that this code is needed in the first place. You should be calling `room.is_available?` here instead of doing all this work yourself. I know we hadn't read chapter 4 of POODR yet when you wrote this, but Metz would say this code is telling how instead of saying what. Alternatively, you could say this is an example of tight coupling.

A loosely coupled version might look something like:

```ruby
def view_available_rooms(date_range)
  return @all_rooms.select do |room|
    room.is_available?(date_range)
  end
end
```

File: spec/concierge_spec.rb [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+    it "loads a list of all rooms" do
+      expect(@concierge.all_rooms).must_be_kind_of Array
+    end
+  end
+  
+  
+  describe "See all rooms method" do
+    before do
+      @concierge = Hotel::Concierge.new
+    end
```
Comment (droberts-sea) [inline spec/concierge_spec.rb]: Since this `before` block is repeated, it might make sense to move it out to the top-most `describe`.

File: spec/concierge_spec.rb [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+  
+  describe "Reserve room method" do
+    before do
+      @concierge = Hotel::Concierge.new
+    end
+    
+    it "selects an available room" do
+     @reservation = @concierge.reserve_room("2019-01-01", "2019-01-03")
+        expect(@reservation.first.room.status).must_equal :AVAILABLE
+    end
```
Comment (droberts-sea) [inline spec/concierge_spec.rb]: This test isn't testing much, since the room statuses are always `:AVAILABLE`.

File: spec/concierge_spec.rb [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+    it "updates the Concierge Reservations array" do
+      @res_count = @concierge.reservations.length
+      @reservation1 = @concierge.reserve_room("2019-01-01", "2019-01-03")
+        expect(@concierge.reservations.length).must_equal (@res_count + 1)
+    end
+    
+    it "updates the Room's list of reservations" do
+      @reservation2 = @concierge.reserve_room("2019-05-01", "2019-05-03")
+    end  
+  end
```
Comment (droberts-sea) [inline spec/concierge_spec.rb:65]: Missing case: what happens if you try to reserve a room, and there are no rooms available for those dates?

File: spec/concierge_spec.rb [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+    res2 = @concierge.reserve_room("2019-3-1", "2019-3-2")
+    res3 = @concierge.reserve_room("2020-1-10", "2020-1-20")
+
+    end
+      it "updates the list of reservations" do
+        expect(@concierge.reservations.length).must_equal 3
+      end
+      
+      it "contains instances of reservations" do
+        expect(@concierge.reservations[2]).must_be_instance_of Hotel::Reservation
```
Comment (droberts-sea) [inline spec/concierge_spec.rb]: Since these two tests don't touch `view_reservations_by_date`, they don't belong here. Maybe in the `describe` block for `reserve_room` instead?

File: spec/room_spec.rb [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+    end  
+    
+    it "updates the room's list of reservations" do   
+      expect(@room.reservations.length).must_equal @count + 1
+      expect(@room.reservations).must_be_kind_of Array
+      expect(@room.reservations[0]).must_be_instance_of Hotel::Reservation
+      expect(@room.reservations[0].id).must_equal 20
+    end
+    
+  end
```
Comment (droberts-sea) [inline spec/room_spec.rb:41]: You should have some tests for `is_available?` here.

**Comment by droberts-sea** [comment]
# Hotel
## What We're Looking For



Feature | Feedback
---     | ---
**Baseline** |
Used git regularly | yes
Answer comprehension questions | yes
**Design** |
Each class is responsible for a single piece of the program | yes
Classes are loosely coupled | In some places this is excellent, and in some it could be improved. `Concierge` in particular is tightly coupled to `Room` and `Reservation`.
**Wave 1** |
List rooms | yes
Reserve a room for a given date range | yes
List reservations for a given date | yes
Calculate reservation price | yes
Invalid date range produces an error | yes
Test coverage | yes
**Wave 2** |
View available rooms for a given date range | close - see inline
Reserving a room that is not available produces an error | close - see inline
Test coverage | some - see inline
**Fundamentals** |
Names variables, classes and modules appropriately | yes
Understanding of variable scope - local vs instance | yes
Can create complex logical structures utilizing variables | some
Appropriately uses methods to break down tasks into smaller simpler tasks | yes
Understands the differences between class and instance methods | yes - one stray class method, but I don't see a pattern
Appropriately uses iterators and `Enumerable` methods | I see a lot of room for improvement here
Appropriately writes and utilizes classes | yes
Appropriately utilizes modules as a namespace | some - see DateRange
**Wrap Up** |
There is a refactors.txt file | nope
The file provides a roadmap to future changes |
**Additional Feedback** |


Hi Amy! I've left a _lot_ of feedback on this one, partly because there's some room for improvement, and partly because this is the first one of these I've done and I'm still fresh. Here is my overall takeaway:


- I am really happy with your high-level design - these classes and their interfaces feel right to me. Good work.
- Wave 1 looks great
- Wave 2 is where you started to run into trouble, particularly with some of the more complex logic for keeping track of reservations and dates
- Your testing follows a similar pattern, starting to break down with the more complex interactions


As you revisit this over the break, that complex logic and the details of the interactions between classes is where I would ask you to focus your efforts.


While there's room for improvement here, I do think that you're in a pretty good place for this point in the course. This stuff is hard, and it takes a lot of practice for it to feel fluid. Keep up the hard work.
